src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Tooltip/DefaultVSLSPTagHelperTooltipFactory.cs

```diff
@@ -33,87 +45,163 @@ internal class DefaultVSLSPTagHelperTooltipFactory : VSLSPTagHelperTooltipFactor
+            AggregateBoundAttributeDescription attributeDescriptionInfo,
+            out List<DescriptionClassification> descriptionClassifications)
         {
-            if (descriptionInfos is null)
-            {
-                throw new ArgumentNullException(nameof(descriptionInfos));
-            }
-
-            var associatedAttributeInfos = descriptionInfos.DescriptionInfos;
+            var associatedAttributeInfos = attributeDescriptionInfo.DescriptionInfos;
             if (associatedAttributeInfos.Count == 0)
             {
-                tagHelperDescription = null;
+                descriptionClassifications = null;
                 return false;
             }
 
+            var descriptions = new List<DescriptionClassification>();
+
             // Generates a ClassifiedTextElement that looks something like:
             //     ReturnType Namespace.TypeName.Property
             //     Summary description
             // with the specific element parts classified appropriately.
-            // Additional entries are separated by two new lines.
 
-            var runs = new List<VSClassifiedTextRun>();
             foreach (var descriptionInfo in associatedAttributeInfos)
             {
-                if (runs.Count > 0)
-                {
-                    runs.Add(NewLine);
-                    runs.Add(NewLine);
-                }
+                // 1. Classify type name and property
+                var typeRuns = new List<VSClassifiedTextRun>();
 
                 if (!TypeNameStringResolver.TryGetSimpleName(descriptionInfo.ReturnTypeName, out var returnTypeName))
                 {
                     returnTypeName = descriptionInfo.ReturnTypeName;
                 }
 
                 var reducedReturnTypeName = ReduceTypeName(returnTypeName);
-                ClassifyReducedTypeName(runs, reducedReturnTypeName);
-                runs.Add(Space);
-                ClassifyTypeName(runs, descriptionInfo.TypeName);
-                runs.Add(Dot);
-                runs.Add(new VSClassifiedTextRun(VSPredefinedClassificationTypeNames.Identifier, descriptionInfo.PropertyName));
-                TryClassifySummary(runs, descriptionInfo.Documentation);
+                ClassifyReducedTypeName(typeRuns, reducedReturnTypeName);
+                typeRuns.Add(Space);
+                ClassifyTypeName(typeRuns, descriptionInfo.TypeName);
+                typeRuns.Add(Dot);
+                typeRuns.Add(new VSClassifiedTextRun(VSPredefinedClassificationTypeNames.Identifier, descriptionInfo.PropertyName));
+
+                // 2. Classify summary
+                var documentationRuns = new List<VSClassifiedTextRun>();
+                TryClassifySummary(documentationRuns, descriptionInfo.Documentation);
+
+                // 3. Combine type + summary information
+                descriptions.Add(new DescriptionClassification(new List<VSClassifiedTextRun>(typeRuns), new List<VSClassifiedTextRun>(documentationRuns)));
```

Comment:
`typeRuns` / `documentationRuns` are re-constructed on every loop. I'd imagine you'd be able to just pass them through here and then not worry about `Clear()`ing them on the two lines below.